### PR TITLE
#19891: ttnn.sort-single-core-implementation

### LINKED ISSUE #19891: Add ttnn.sort basic implementation
Building on #19548 keep implementing ttnn.sort based on the definition found here:
https://pytorch.org/docs/stable/generated/torch.sort.html

In addition to returning the correct shape, the operation should now return the correct data along the given dimension.  Consider starting with a few small tensors in the examples from pytorch.org listed above, or an example below, and verify result when sorting on dim=0, dim=1, and dim=2.
```
x = torch.tensor([
    [[6, 3, 9, 1],
     [5, 2, 8, 4],
     [7, 0, 3, 2]],

    [[1, 7, 4, 6],
     [8, 5, 2, 3],
     [0, 9, 6, 1]]
])
```

Focus on a single core implementation, and avoid large tensors so that the working set fits in L1 cache.  After the trivial tensor sort operations are working correctly, try tile-sized tensor shapes.

### PR DESCRIPTION
**Pull request by mgajewskiTT**
### Ticket
Add ttnn.sort basic implementation - [#19891](https://github.com/tenstorrent/tt-metal/issues/19891)

### Problem description
Single core implementation on `ttnny.experimental.sort` with the use of ` ckernel::topk_local_sort` LLK.

### What's changed
- Added tensor sorting on single core,
- Updated tests.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp [hunk 1]
```diff
@@ -7,17 +7,69 @@
...
+    const bool is_rank_le_4d,
+    const bool descending) {
+    if (input_tensor.get_logical_shape() == ttnn::Shape{1}) {
+        // Early exit for scalar tensors, return the same tensor
+        // Scalar tensors do not require sorting.
+        return input_tensor;
+    }
+    // If dim is not last dimension transpose it
+    const Tensor transposed_tensor = reduction_common::perform_transpose(input_tensor, is_dim_last_idx, dim, -1);
+    // If input is not rank 4 transorm it to 4D
```
[the remark this answers is not in the document]
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp:34]: I've tested it on tensors dim 5 and above and it works correctly. The LLK only sorts in one dimension, and it does not have option to choose which one it is, so we transpose input tensor here so that the sorting dimension will always be the last one.  I’ll check the option to avoid transforming to and from 4D, so far I based my implementation on topk, which uses a similar mechanism, and this transformation also makes it easier to introduce parallelism, which I am currently working on.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/compute/sort_common.hpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+namespace NAMESPACE {
+
+void sort_Wt_tiles_row_to_bitonic_sequence(
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/compute/sort_common.hpp:7]: Since this is all done up front, does that limit how large the input tensors can be?

Also, if you don't do this is this an easy path to sort by other dimensions?

And would transposing to the second last dimension work better (and avoid the limit on the input tensor size)?

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/reader.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    cb_reserve_back(cb_id, 1);
+
+    // Writer config
+    const uint32_t writer_addr = get_write_ptr(cb_id);
+    volatile tt_l1_ptr uint16_t* ptr = reinterpret_cast<volatile tt_l1_ptr uint16_t*>(writer_addr);
+    const uint16_t wt_offset = wt << 5;  // wt * 2^(5)
+
+    // Writer loop
+    uint32_t count = 0;
+    for (uint32_t i = 0; i < 2; ++i) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/reader.cpp:21]: Please add a comment explaining why this loop is structured this way. Also, please create constexpr variables for the 2s and 16s.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/reader.cpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    for (uint32_t h = 0; h < Ht; h++) {
+        for (uint32_t w = 0; w < Wt; w++) {
+            cb_reserve_back(input_tensor_cb_index, one_tile);
+            const uint32_t l1_write_addr = get_write_ptr(input_tensor_cb_index);
+            noc_async_read_tile(h * Wt + w, s, l1_write_addr);
+            noc_async_read_barrier();
+            cb_push_back(input_tensor_cb_index, one_tile);
+
+            generate_index_tile(index_tensor_cb_index, w);
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/reader.cpp:66]: The reader / writer kernels are somewhat interchangable. For performance, a lot of ops split the work across the reader / writer.

Here it would probably make sense to have the writer generate the index tiles.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/writer.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        for (uint32_t w = 0; w < Wt; w++) {
+            cb_wait_front(value_tensor_cb_index, one_tile);
+            const uint32_t l1_write_addr_val = get_read_ptr(value_tensor_cb_index);
+            noc_async_write_tile(h * Wt + w, interleaved_accessor0, l1_write_addr_val);
+            noc_async_write_barrier();
+            cb_pop_front(value_tensor_cb_index, one_tile);
+        }
+
+        // Index tensor
+        for (uint32_t w = 0; w < Wt; w++) {
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/writer.cpp:51]: The reader / writer kernels are somewhat interchangable. For performance, a lot of ops split the work across the reader / writer.

Here it would probably make sense to have the reader write back either the value or index tensor.

With the changes, it would be good to profile the perf.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/compute/sort.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        cb_wait_front(input_tensor_transposed_cb_index, Wt);
+        cb_wait_front(index_tensor_transposed_cb_index, Wt);
+
+        // Sort and merge step of bitonic merge sort
+        uint32_t stages = 0;
+        for (uint32_t temp = Wt; temp > 1; temp >>= 1) {
+            stages++;
+        }
+
+        for (uint32_t stage = 2; stage <= stages; stage++) {
```
[the remark this answers is not in the document]
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/compute/sort.cpp:99]: I added full algorithm description on top of the file.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/sort_device_operation.cpp [hunk 7]
```diff
@@ -3,26 +3,59 @@
...
+        "Operands need to be allocated in buffers on the device. Buffer is null.");
+    TT_FATAL(
+        tensor_args.input_tensor.storage_type() == StorageType::DEVICE,
+        "Operation requires input to be on Device. Input storage type: {}",
+        static_cast<int>(tensor_args.input_tensor.storage_type()));
+
+    TT_FATAL(input_tensor_shape.rank() == 4, "Input shape must be 4D, got {}", input_tensor_shape.rank());
+
+    TT_FATAL(
+        input_tensor_shape[-1] % 64 == 0,
```
[the remark this answers is not in the document]
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/sort_device_operation.cpp:37]: I added information about this to pybind where the OP docs are. 

I was thinking about manually adding padding to resolve this in the future, for now I was thinking about focusing on single core implementation, then multicore implementation then in the last stage resolving limitations.

File: tests/ttnn/unit_tests/operations/reduce/test_sort.py [hunk 8]
```diff
@@ -34,31 +30,33 @@ def test_sort_output_shape(shape, dim, descending, device):
...
 
     assert torch_sort_values.shape == ttnn_sort_values.shape
-    assert torch_sort_indeces.shape == ttnn_sort_indices.shape
+    assert torch_sort_indices.shape == ttnn_sort_indices.shape
 
     assert list(ttnn_sort_values.shape) == shape
     assert list(ttnn_sort_indices.shape) == shape
 
+    if len(shape) == 0 or len(shape) == 1:
+        torch_sort_values == ttnn.to_torch(ttnn_sort_values)
```
[the remark this answers is not in the document]
Comment (mgajewskiTT) [reply tests/ttnn/unit_tests/operations/reduce/test_sort.py]: I missed that. Added.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/writer.cpp (hunk 5 above)
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/writer.cpp]: I changed both reader and writer accordingly.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/reader.cpp (hunk 3 above)
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/dataflow/reader.cpp]: Added.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp:34]: Can you please provide a specific example?

[the remark this answers is not in the document]
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp:34]: I wrote a simple test:
```python
def test_tensor(device):
    torch_dtype = torch.bfloat16
    
    input = torch.randn([32, 32, 32, 32, 64], dtype=torch_dtype)
    
    print(f"1. Input shape: {input.shape}")
    print(f"2. Tensor rank: {input.ndim}")

    torch_sort_values, torch_sort_indices = torch.sort(input, dim=-1, descending=False)

    ttnn_input = ttnn.from_torch(input, ttnn.bfloat16, layout=ttnn.Layout.TILE, device=device)
    ttnn_sort_values, ttnn_sort_indices = ttnn.experimental.sort(ttnn_input, dim=-1)
    
    assert_with_pcc(torch_sort_values, ttnn.to_torch(ttnn_sort_values))
```
It passes.

[the remark this answers is not in the document]
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp:34]: Does this test work if you don't do the transform_to_4d_tensor / squeeze_from_4D / reshape?

If the test doesn't work, how hard would it be to make it work?

Also, for this test, how does the transform_to_4d_tensor / squeeze_from_4D / reshape help with parallelism?

[the remark this answers is not in the document]
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp:34]: Unfortunately, it doesn't work, in many places throughout the program the sizes are calculated and conditions are checked with respect to the tensor, which has four dimensions. Changing and adjusting everything again would take some time. Additionally, wouldn't the lack of data transformation to 4D completely change the arrangement of data in individual tiles? Then the kernels would also have to be changed, my idea of this transformation, apart from basing it on the topK implementation, was to create the processing of one data format and adjusting it at the beginning and exit of the function – makes it universal for all the cases.

Regarding parallelization, according to the description of the task https://github.com/tenstorrent/tt-metal/issues/20623 parallelization takes place on dimensions other than the one selected by the user. I even managed to create a first version of this approach: https://github.com/tenstorrent/tt-metal/blob/mgajewski/20623-add-ttnnsort-multi-core-implementation/ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/kernels/compute/sort_multicore.cpp Wouldn't the lack of data transformation to 4D change the data arrangement in the tiles and the lack of a clear division into Wt and Ht?

I might be wrong then please correct me.

[the remark this answers is not in the document]
Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp:34]: The tensor gets mapped to 2D in memory for tiles, and there is always a clear Wt and Ht. Changing dimensions should be a change in view. However, that is also probably low cost.

For now it's okay to keep the existing functionality. However, once you look at performance you might need to revisit this if the overhead is too high.

**Review by bbradelTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp [hunk 9]
```diff
@@ -55,7 +55,7 @@ Tensor pre_sort_transform_tensor(
     const Tensor transformed_tensor = reduction_common::transform_to_4d_tensor(transposed_tensor, is_rank_le_4d);
     // Fill implicit tile padding with the appropriate value
     const Tensor padded_tensor = ttnn::fill_implicit_tile_padding(
-        transformed_tensor, descending ? std::numeric_limits<float>::min() : std::numeric_limits<float>::max());
+        transformed_tensor, descending ? std::numeric_limits<float>::lowest() : std::numeric_limits<float>::max());
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp:58]: What happens if there are positive or negative infinities in the input?

Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/sort.cpp]: Good point, changed to `-std::numeric_limits<float>::infinity()`. I noticed that ::min() returns absolute minimum number instead of actual minimum available for float so for tensors with negative values there were mistakes. All fixed now.

File: ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/sort_device_operation.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mgajewskiTT) [reply ttnn/cpp/ttnn/operations/experimental/reduction/sort/device/sort_device_operation.cpp:37]: This is no longer an issue, since algorithm works on number of tile that need to be power of 2, I added padding mechanism that also solves this problem. I leave this check here just for safety.

**Review by aczajkowskiTT — APPROVED** [review]
Based on our code review the OP performs as intended

**Review by nsorabaTT — APPROVED** [review]

**Review by edwinleeTT — APPROVED** [review]

**Review by TT-BrianLiu — COMMENTED** [review]
General comment about the op infra. This seems to be using the new TMP version @ayerofieiev-tt What are your thoughts on this? I thought eventually we will move away from this

**Review by vsureshTT — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]
